src/java/org/apache/cassandra/streaming/messages/OutgoingMutationLogStreamMessage.java

```diff
@@ -0,0 +1,135 @@
+
+    public static final StreamMessage.Serializer<OutgoingMutationLogStreamMessage> serializer = new OutgoingMutationLogStreamMessageSerializer();
+
+    private final MutationJournal.Snapshot snapshot;
+
+    public OutgoingMutationLogStreamMessage(LogStreamHeader header, MutationJournal.Snapshot snapshot)
+    {
+        super(header);
+        this.snapshot = snapshot;
+    }
+
+    public void serialize(StreamingDataOutputPlus out, int version, StreamSession session) throws IOException
+    {
+        LogStreamHeader.serializer.serialize(header, out, version);
+
+        try
+        {
+            // Stream mutations using the journal readAll method and filter by keyspace and token ranges
+            snapshot.readAll((segment, position, key, buffer, userVersion) -> {
+                try (DataInputBuffer in = new DataInputBuffer(buffer, true))
+                {
+                    Pair<DecoratedKey, TableMetadata> keyAndTableMetadata = Mutation.serializer.deserializeKeyAndTableMetadata(in, userVersion, DeserializationHelper.Flag.LOCAL);
+                    DecoratedKey dk = keyAndTableMetadata.left;
+                    String keyspace = keyAndTableMetadata.right.keyspace;
+
+                    // don't send fully reconciled mutations
+                    if (header.reconciled.isFullyReconciled(keyspace, key))
+                        return;
+
+                    // Check if the mutation's keyspace and token are in our ranges
+                    Set<Range<Token>> ranges = header.manifest.keyspaceRanges.get(keyspace);
+
+                    if (ranges == null)
+                    {
+                        if (logger.isTraceEnabled())
+                            logger.trace("Mutation {} not sent: keyspace {} not in manifest ranges for session {}", key, keyspace, session.planId());
+                        return;
+                    }
+
+                    if (!Range.isInRanges(dk.getToken(), ranges))
+                    {
+                        if (logger.isTraceEnabled())
+                            logger.trace("Mutation {} not sent: token {} not in ranges for keyspace {} in session {}", key, dk.getToken(), keyspace, session.planId());
+                        return;
+                    }
+
+                    if (logger.isTraceEnabled())
+                        logger.trace("Sending mutation {}: keyspace={}, token={}, session={}", key, keyspace, dk.getToken(), session.planId());
+
+                    out.writeBoolean(true);
+                    out.writeInt(userVersion);
+                    ByteBufferUtil.writeWithVIntLength(buffer, out);
+                }
+                catch (IOException e)
+                {
+                    throw new RuntimeException(e);
+                }
+            });
+        }
+        finally
+        {
+            snapshot.close();
+        }
+
+        // end-of-stream marker
+        out.writeBoolean(false);
+
+        session.logStreamSent(this);
```

Comment:
Should we avoid side effects like timeout scheduling during serde?

Reply:
this is the same pattern used by sstable streaming